### ISSUE TITLE
[Bug]: `react/no-array-index-key` falsely triggered for `Map.prototype.forEach()`

### ISSUE THREAD

**Issue by RulerOfCakes** [jsx-eslint/eslint-plugin-react#3670]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

When using the `forEach` method on a Map, the second parameter given to the callback function is not an index, but rather a key of the map. I believe this is quite different from a typical array index and doesn't align with the reasons why `react/no-array-index-key` exists, and it isn't really displayed in the [example document](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-array-index-key.md) as well.

For example, some code like this triggers the warning:

```javascript
const map = new Map();
...
return (<div>
  {map.forEach((item, key)=><div key={key}>{item}</div>)}
</div>)
```
Whereas I think it is reasonable to assume the key given here is a unique ID that is different from a typical array index, since it is a map, therefore giving it a warning by the name `no-array-index-key` is quite unintuitive.

### Expected Behavior

With the same code above, there should be no warning regarding the key props inserted in the return statement of the `Map.prototype.forEach()` call, or at the very least a warning with a different description to not give it an array-specific warning title.

### eslint-plugin-react version

v7.28.0

### eslint version

v8.2.0

### node version

v18.15.0

**Comment by ljharb** [comment]
You’re right except that forEach always returns undefined, so this won’t actually render anything at all - iow, it’s a bug that you’re lucky enough the lint rule caught. You’ll have to use Array.from and a mapper function to do it properly, at which point the index rule would apply.

**Comment by RulerOfCakes** [comment]
> You’re right except that forEach always returns undefined, so this won’t actually render anything at all - iow, it’s a bug that you’re lucky enough the lint rule caught. You’ll have to use Array.from and a mapper function to do it properly, at which point the index rule would apply.

Wow, sorry I had completely missed that part. I will rephrase the issue again, as I believe the false alarm still applies in situations like this:
```javascript
const map = new Map();
...
const results = [];
// react/no-array-index-key triggered here
map.forEach((item, key) => results.push(<div key={key}>{item}</div>));

return <div>{results.map((item) => item)}</div>;
```

Although I admit this is a bit of a niche use case...

**Comment by ljharb** [comment]
It is; forEach is for side effects, side effects aren’t appropriate in React (except with useEffect), and your use case calls for .map :-)
